src/fetch_metrics.py

```diff
@@ -0,0 +1,250 @@
+        print(f"Error reading label descriptions file: {e}")
+        return {}
+
+
+# Retry logic for HTTP requests
+def make_request_with_retries(url, params=None, headers=None, verify=True, retries=3, timeout=30):
+    for attempt in range(retries):
+        try:
+            response = requests.get(url, params=params, headers=headers, verify=verify, timeout=timeout)
+            response.raise_for_status()
+            return response
+        except requests.exceptions.RequestException as e:
+            print(f"Attempt {attempt + 1} failed: {e}")
+            if attempt == retries - 1:
+                raise
+
+
+# Fetch all metric names filtered by prefix
+def fetch_metric_names_by_prefix(prefix):
+    try:
+        response = make_request_with_retries(
+            f"{prometheus_url}/api/v1/label/__name__/values",
+            headers=headers,
+            verify=ca_cert_path or False
+        )
+        all_metrics = response.json().get('data', [])
+        return [metric for metric in all_metrics if metric.startswith(prefix)]
+    except requests.exceptions.RequestException as e:
+        print(f"Error fetching metric names with prefix '{prefix}': {e}")
+        return []
+
+
+# Fetch metadata for a specific metric, including description, labels, and type
+def fetch_metric_metadata(metric_name):
+    try:
+        # Fetch metadata for the specific metric
+        metadata_response = make_request_with_retries(f"{prometheus_url}/api/v1/metadata", headers=headers, verify=ca_cert_path or False)
+        all_metadata = metadata_response.json().get('data', {})
+        metadata = all_metadata.get(metric_name, [])
+
+        # Extract description and type
+        metric_description = "No description available"
+        metric_type = "Unknown"
+        for meta in metadata:
+            if 'help' in meta:
+                metric_description = meta['help']
+            if 'type' in meta:
+                metric_type = meta['type']
+                break
+
+        # Fetch labels
+        labels_response = make_request_with_retries(f"{prometheus_url}/api/v1/query", params={'query': metric_name}, headers=headers, verify=ca_cert_path or False)
+        results = labels_response.json().get('data', {}).get('result', [])
+
+        # Extract labels
+        labels = set()
+        for result in results:
+            labels.update(result.get('metric', {}).keys())
+
+        # Create a structured data format for YAML
+        return {
+            'metric_name': metric_name,
+            'metric_description': metric_description,
+            'type': metric_type,
+            'labels': [{'label_name': label, 'label_description': 'Description not available'} for label in labels]
+        }
+
+    except requests.exceptions.RequestException as e:
+        print(f"Error fetching metadata for metric '{metric_name}': {e}")
+        return {
+            'metric_name': metric_name,
+            'metric_description': "No description available",
```

Comment:
There is `No description available` phrase here, while `Description not available` in the labels. It would be better to standardize the way the unavailable data is provided. I would actually vote for `null`-like value, as it makes it easier on the client side to decide how to deal with missing data.

Reply:
Thank you! I will update

Reply:
Done